src/groups/mqb/mqbblp/mqbblp_cluster.cpp

```diff
@@ -3674,5 +3716,118 @@ void Cluster::processResponse(const bmqp_ctrlmsg::ControlMessage& response)
         this);
 }
 
+void Cluster::getPrimaryNodes(bsl::vector<mqbnet::ClusterNode*>* nodes,
+                              bool*                              isSelfPrimary,
+                              mqbcmd::InternalResult*            result) const
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(nodes);
+    BSLS_ASSERT_SAFE(isSelfPrimary);
+    BSLS_ASSERT_SAFE(result);
+    BSLS_ASSERT_SAFE(dispatcher()->inDispatcherThread(this));
+
+    const mqbc::ClusterState::PartitionsInfo& partitionsInfo =
+        d_state.partitions();
+
+    nodes->clear();
+    *isSelfPrimary = false;
+
+    for (mqbc::ClusterState::PartitionsInfo::const_iterator pit =
+             partitionsInfo.begin();
+         pit != partitionsInfo.end();
+         pit++) {
+        if (pit->primaryStatus() !=
+            // TODO: Handle this case (will want to buffer)
+            bmqp_ctrlmsg::PrimaryStatus::Value::E_ACTIVE) {
+            BALL_LOG_WARN << "While collecting primary nodes: "
+                          << "primary for partition " << pit->partitionId()
+                          << " is not active";
+            result->makeError().message() =
+                "Primary is not active for partition id " +
+                bsl::to_string(pit->partitionId());
+            return;
+        }
+
+        mqbnet::ClusterNode* primary = pit->primaryNode();
+
+        if (primary) {
+            // Don't add duplicate
+            if (bsl::find(nodes->begin(), nodes->end(), primary) !=
+                nodes->end()) {
+                continue;
+            }
+            // Check for self
+            if (d_state.isSelfActivePrimary(pit->partitionId())) {
+                *isSelfPrimary = true;
+                continue;
+            }
+            nodes->push_back(primary);
+        }
+        else {
+            BALL_LOG_WARN << "Error while collecting primary nodes: No "
+                             "primary found for partition id "
+                          << pit->partitionId();
+            result->makeError().message() =
+                "No primary found for partition id " +
+                bsl::to_string(pit->partitionId());
+            return;
+        }
+    }
+}
+
+void Cluster::getPartitionPrimaryNode(mqbnet::ClusterNode**   node,
+                                      bool*                   isSelfPrimary,
+                                      mqbcmd::InternalResult* result,
+                                      int partitionId) const
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(node);
+    BSLS_ASSERT_SAFE(isSelfPrimary);
+    BSLS_ASSERT_SAFE(result);
+    BSLS_ASSERT_SAFE(dispatcher()->inDispatcherThread(this));
```

Comment:
Since the `partitionId` can originate from someone executing `CLUSTERS CLUSTER <name> STORAGE PARTITION <partitionId> ENABLE`, for example, I prefer to send an error message in this case instead of using an assert.

Reply:
Agree